### DISCUSSION TITLE
Can you use variables in translated strings?

### DISCUSSION THREAD

**Discussion by Timo654** [MyreMylar/pygame_gui#261] [Q&A] [answered]
I'm trying to translate my program and some of the text has variables in it (stuff like names and a volume slider), but I couldn't figure out how to make the variables work.

Is there a way to do this?

Thanks!

**Comment by MyreMylar** [comment]
Hello, the translation library use in pygame GUI supports this with
placeholders in translated strings. See their documentation here:

https://github.com/danhper/python-i18n#placeholder

Hopefully that helps, on my phone on the go at the moment.

**Reply by Timo654** [reply]
Hi,

Thanks for that. I'm not sure how I can use that through pygame_gui, but this did kinda work:
```python
volume_label = UILabel(pygame.Rect((215, 402), (-1, 25)),
                           i18n.t('kuma_ui.volume_label', volume=volume_value)),  
                           manager=manager)
```
The line in my translation file:
```json
"volume_label": "Volume %{volume}"
```

It does work, but the only issue is that the text is not going to update when switching the locale. Is there a better way of doing this? I couldn't figure out how I could pass more parameters to the "text" param.

**Comment by MyreMylar** [comment, marked as answer]
Hello, I had a look into this and it was pretty annoying to add parameters to a UI Label's translation. I think I just used the functionality once in the File Loading/Saving dialog - but there I had complete control over what I was translating and the dialog wasn't really a plausible place to switch locale in an application.

I've made a quick trial patch to the main branch (what will be 0.6.5 at some point) which adds optional kwargs to a `UILabel` in the constructor/init and in the `set_text()` function. These are then passed through to the translate function so the interface would be something like:

Code:
```
test_label = pygame_gui.elements.UILabel(pygame.Rect((10, 100), (150, 30)),
                                         'examples.label_test',
                                         manager,
                                         a_number=str(10))
```

Translation file:
```
    "label_test": "Testing a label plus %{a_number}"
```

If you know how to grab a repository from github and build it yourself with ` pip install . -U` then you could give it a try and see if it works for your use case. 

I'll probably roll out a 0.6.5 version over the Easter period when I have a break from work with whatever other issues have come up before  that time period.

**Reply by Timo654** [reply]
Hey,

Thanks, I just tested it. It seems to work, however when using the `set_text()` function, the label will not get rebuilt.
I fixed it by adding `any_changed = True` after line 94 and line 96 in `pygame_gui/elements/ui_label.py`.

Other than that, it seems to work fine so far.

EDIT: Would it be possible to also support this for other elements, such as UIMessageWindow and UIConfirmationDialog?

**Comment by MyreMylar** [comment]
Thanks for testing it! I will fix that set_text() bug before release.

**Reply by Timo654** [reply]
No problem.

In case you missed it, would it be possible to also support this for other elements, such as UIMessageWindow and UIConfirmationDialog? I tried using i18n directly for those, but it didn't update the text when I changed language and tried opening the same window again.

Thanks!

**Reply by MyreMylar** [reply]
Yeah, I want to double check that what I've done with the label is sensible over the Easter break, but assuming all is well I'll extend the same methodology to all the places where you can enter a custom string that is passed to the translation software - like the UIMessageWindow and the UIConfirmationDialog.

**Reply by Timo654** [reply]
> Yeah, I want to double check that what I've done with the label is sensible over the Easter break, but assuming all is well I'll extend the same methodology to all the places where you can enter a custom string that is passed to the translation software - like the UIMessageWindow and the UIConfirmationDialog.

Any news on this?

**Comment by MyreMylar** [comment]
I haven't had a chance to look at it yet sorry.

I should probably open an issue on the tracker to remind myself too.
